File: backend/app/models/product.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel

ProductCategory = Literal["SHIRT", "ITEM"]


class Product(BaseModel):
    """Mirrors a row in the Products sheet.

    Shirts (variant_required=True) track stock per size in Product_Variants;
    `stock` on the product row itself is unused for those and should be
    ignored in favor of the sum of variant stocks.
    """

    product_id: str
    name: str
    description: str
    category: ProductCategory
    price: float
    stock: int
    active: bool = True
    image_url: str = ""
    variant_required: bool = False
    created_at: datetime
    updated_at: datetime


class ProductVariant(BaseModel):
    """Mirrors a row in the Product_Variants sheet (e.g. a shirt size)."""

    product_id: str
    variant: str
    stock: int

File: frontend/types/product.ts
export type ProductCategory = "SHIRT" | "ITEM";

export interface ProductVariant {
  variant: string;
  stock: number;
}

export interface Product {
  product_id: string;
  name: string;
  description: string;
  category: ProductCategory;
  price: number;
  stock: number;
  active: boolean;
  image_url: string;
  variant_required: boolean;
  variants: ProductVariant[];
}

export interface ProductCreateRequest {
  product_id: string;
  name: string;
  description?: string;
  category: ProductCategory;
  price: number;
  stock?: number;
  image_url?: string;
  variant_required?: boolean;
  variants?: ProductVariant[];
}

export interface ProductUpdateRequest {
  name?: string;
  description?: string;
  category?: ProductCategory;
  price?: number;
  stock?: number;
  active?: boolean;
  image_url?: string;
  variant_required?: boolean;
}

export interface VariantUpsertRequest {
  variant: string;
  stock: number;
}
